File: hellokratos/internal/service/referral.go
package service

import (
	"context"
	"fmt"

	"google.golang.org/protobuf/types/known/timestamppb"

	pb "hellokratos/api/rider/v1"
	"hellokratos/internal/biz"
	"hellokratos/internal/data/model"
)

// ReferralService 拉新服务
type ReferralService struct {
	pb.UnimplementedReferralServer
	uc *biz.ReferralUsecase
}

// NewReferralService 创建拉新服务
func NewReferralService(uc *biz.ReferralUsecase) *ReferralService {
	return &ReferralService{
		uc: uc,
	}
}

// GenerateInviteCode 生成邀请码
func (s *ReferralService) GenerateInviteCode(ctx context.Context, req *pb.GenerateInviteCodeRequest) (*pb.GenerateInviteCodeReply, error) {
	inviteCode, err := s.uc.GenerateInviteCode(ctx, req.RiderId)
	if err != nil {
		return nil, fmt.Errorf("生成邀请码失败: %w", err)
	}

	return &pb.GenerateInviteCodeReply{
		InviteCodeInfo: &pb.InviteCodeInfo{
			InviteCode:   inviteCode.InviteCode,
			InviteLink:   inviteCode.InviteLink,
			CreatedAt:    timestamppb.New(inviteCode.CreatedAt),
			ExpireAt:     timestamppb.New(inviteCode.ExpireAt),
			TotalInvited: inviteCode.TotalInvited,
			ValidInvited: inviteCode.ValidInvited,
			TotalRewards: inviteCode.TotalRewards,
		},
	}, nil
}

// GetMyReferralInfo 获取我的邀请信息
func (s *ReferralService) GetMyReferralInfo(ctx context.Context, req *pb.GetMyReferralInfoRequest) (*pb.GetMyReferralInfoReply, error) {
	inviteCode, todayInvited, monthInvited, pendingRewards, err := s.uc.GetMyReferralInfo(ctx, req.RiderId)
	if err != nil {
		return nil, fmt.Errorf("获取邀请信息失败: %w", err)
	}

	reply := &pb.GetMyReferralInfoReply{
		TodayInvited:   todayInvited,
		MonthInvited:   monthInvited,
		PendingRewards: int32(pendingRewards),
	}

	if inviteCode != nil {
		reply.InviteCodeInfo = &pb.InviteCodeInfo{
			InviteCode:   inviteCode.InviteCode,
			InviteLink:   inviteCode.InviteLink,
			CreatedAt:    timestamppb.New(inviteCode.CreatedAt),
			ExpireAt:     timestamppb.New(inviteCode.ExpireAt),
			TotalInvited: inviteCode.TotalInvited,
			ValidInvited: inviteCode.ValidInvited,
			TotalRewards: inviteCode.TotalRewards,
		}
	}

	return reply, nil
}

// GetInviteRecordList 获取邀请记录列表
func (s *ReferralService) GetInviteRecordList(ctx context.Context, req *pb.GetInviteRecordListRequest) (*pb.GetInviteRecordListReply, error) {
	if req.Page <= 0 {
		req.Page = 1
	}
	if req.PageSize <= 0 {
		req.PageSize = 10
	}

	relations, total, err := s.uc.GetInviteRecordList(ctx, req.RiderId, req.Status, int(req.Page), int(req.PageSize))
	if err != nil {
		return nil, fmt.Errorf("获取邀请记录失败: %w", err)
	}

	records := make([]*pb.InvitedRiderInfo, 0, len(relations))
	for _, relation := range relations {
		info := &pb.InvitedRiderInfo{
			RiderId: relation.InvitedID,
			Status:  relation.Status,
		}
		if relation.RegisterAt.Unix() > 0 {
			info.RegisterAt = timestamppb.New(relation.RegisterAt)
		}
		if relation.FirstOrderAt != nil {
			info.FirstOrderAt = timestamppb.New(*relation.FirstOrderAt)
		}
		if relation.TaskCompletedAt != nil {
			info.TaskCompletedAt = timestamppb.New(*relation.TaskCompletedAt)
		}
		info.RewardAmount = relation.RewardAmount
		records = append(records, info)
	}

	return &pb.GetInviteRecordListReply{
		Records:  records,
		Total:    int32(total),
		Page:     req.Page,
		PageSize: req.PageSize,
	}, nil
}

// ValidateInviteCode 验证邀请码
func (s *ReferralService) ValidateInviteCode(ctx context.Context, req *pb.ValidateInviteCodeRequest) (*pb.ValidateInviteCodeReply, error) {
	isValid, inviterName, err := s.uc.ValidateInviteCode(ctx, req.InviteCode)
	if err != nil {
		return nil, fmt.Errorf("验证邀请码失败: %w", err)
	}

	message := ""
	if !isValid {
		message = "邀请码无效或已过期"
	} else {
		message = "邀请码有效"
	}

	return &pb.ValidateInviteCodeReply{
		IsValid:     isValid,
		InviterName: inviterName,
		Message:     message,
	}, nil
}

// BindReferralRelation 绑定邀请关系
func (s *ReferralService) BindReferralRelation(ctx context.Context, req *pb.BindReferralRelationRequest) (*pb.BindReferralRelationReply, error) {
	inviterID, err := s.uc.BindReferralRelation(ctx, req.NewRiderId, req.InviteCode, req.Phone)
	if err != nil {
		return &pb.BindReferralRelationReply{
			Success: false,
			Message: err.Error(),
		}, nil
	}

	return &pb.BindReferralRelationReply{
		Success:   true,
		Message:   "绑定成功",
		InviterId: inviterID,
	}, nil
}

// GetTaskList 获取任务列表
func (s *ReferralService) GetTaskList(ctx context.Context, req *pb.GetTaskListRequest) (*pb.GetTaskListReply, error) {
	tasks, err := s.uc.GetTaskList(ctx)
	if err != nil {
		return nil, fmt.Errorf("获取任务列表失败: %w", err)
	}

	taskInfos := make([]*pb.TaskInfo, 0, len(tasks))
	for _, task := range tasks {
		taskInfos = append(taskInfos, &pb.TaskInfo{
			TaskId:        task.ID,
			TaskName:      task.TaskName,
			TaskDesc:      task.TaskDesc,
			TaskType:      task.TaskType,
			TargetValue:   task.TargetValue,
			RewardAmount:  task.RewardAmount,
			TimeLimitDays: task.TimeLimitDays,
			StartTime:     timestamppb.New(task.StartTime),
			EndTime:       timestamppb.New(task.EndTime),
		})
	}

	return &pb.GetTaskListReply{
		Tasks: taskInfos,
	}, nil
}

// GetTaskProgress 获取任务进度
func (s *ReferralService) GetTaskProgress(ctx context.Context, req *pb.GetTaskProgressRequest) (*pb.GetTaskProgressReply, error) {
	progressList, err := s.uc.GetTaskProgress(ctx, req.RiderId)
	if err != nil {
		return nil, fmt.Errorf("获取任务进度失败: %w", err)
	}

	// 查找指定任务的进度
	var targetProgress *model.ReferralTaskProgress
	for _, p := range progressList {
		if p.TaskID == req.TaskId {
			targetProgress = p
			break
		}
	}

	if targetProgress == nil {
		return nil, fmt.Errorf("任务不存在")
	}

	// 计算进度百分比
	progressPercent := int32(0)
	if targetProgress.TargetValue > 0 {
		progressPercent = targetProgress.CurrentValue * 100 / targetProgress.TargetValue
	}

	return &pb.GetTaskProgressReply{
		Progress: &pb.TaskProgressInfo{
			TaskId:          targetProgress.TaskID,
			CurrentValue:    targetProgress.CurrentValue,
			TargetValue:     targetProgress.TargetValue,
			ProgressPercent: progressPercent,
			Status:          targetProgress.Status,
			Deadline:        timestamppb.New(targetProgress.Deadline),
			IsClaimed:       targetProgress.IsClaimed,
		},
	}, nil
}

// ClaimTaskReward 领取任务奖励
func (s *ReferralService) ClaimTaskReward(ctx context.Context, req *pb.ClaimTaskRewardRequest) (*pb.ClaimTaskRewardReply, error) {
	rewardAmount, err := s.uc.ClaimTaskReward(ctx, req.RiderId, req.TaskId)
	if err != nil {
		return &pb.ClaimTaskRewardReply{
			Success: false,
			Message: err.Error(),
		}, nil
	}

	return &pb.ClaimTaskRewardReply{
		Success:      true,
		RewardAmount: rewardAmount,
		Message:      "领取成功",
	}, nil
}

// GetRewardRecordList 获取奖励记录列表
func (s *ReferralService) GetRewardRecordList(ctx context.Context, req *pb.GetRewardRecordListRequest) (*pb.GetRewardRecordListReply, error) {
	if req.Page <= 0 {
		req.Page = 1
	}
	if req.PageSize <= 0 {
		req.PageSize = 10
	}

	records, total, totalAmount, err := s.uc.GetRewardRecordList(ctx, req.RiderId, req.Status, int(req.Page), int(req.PageSize))
	if err != nil {
		return nil, fmt.Errorf("获取奖励记录失败: %w", err)
	}

	recordList := make([]*pb.RewardRecord, 0, len(records))
	for _, record := range records {
		r := &pb.RewardRecord{
			RecordId:     record.ID,
			TaskId:       record.TaskID,
			RewardAmount: record.RewardAmount,
			RewardType:   record.RewardType,
			Status:       record.Status,
			CreatedAt:    timestamppb.New(record.CreatedAt),
		}
		if record.IssuedAt != nil {
			r.IssuedAt = timestamppb.New(*record.IssuedAt)
		}
		recordList = append(recordList, r)
	}

	return &pb.GetRewardRecordListReply{
		Records:     recordList,
		Total:       int32(total),
		TotalAmount: int32(totalAmount),
	}, nil
}

// GetReferralStatistics 获取拉新统计
func (s *ReferralService) GetReferralStatistics(ctx context.Context, req *pb.GetReferralStatisticsRequest) (*pb.GetReferralStatisticsReply, error) {
	stats, err := s.uc.GetReferralStatistics(ctx, req.StartDate, req.EndDate)
	if err != nil {
		return nil, fmt.Errorf("获取统计数据失败: %w", err)
	}

	var totalInvited, validInvited, totalRewards int32
	dailyStats := make([]*pb.DailyStatistics, 0, len(stats))

	for _, stat := range stats {
		totalInvited += stat.NewInvited
		validInvited += stat.ValidInvited
		totalRewards += stat.RewardsAmount

		dailyStats = append(dailyStats, &pb.DailyStatistics{
			Date:          stat.StatDate,
			NewInvited:    stat.NewInvited,
			ValidInvited:  stat.ValidInvited,
			RewardsAmount: stat.RewardsAmount,
		})
	}

	// 计算转化率
	conversionRate := int32(0)
	if totalInvited > 0 {
		conversionRate = validInvited * 100 / totalInvited
	}

	return &pb.GetReferralStatisticsReply{
		TotalInvited:   totalInvited,
		ValidInvited:   validInvited,
		TotalRewards:   totalRewards,
		ConversionRate: conversionRate,
		DailyStats:     dailyStats,
	}, nil
}

File: rider-app/src/api/referral.ts
import apiClient from './client';

export interface InviteCodeInfo {
  invite_code: string;
  invite_link: string;
  created_at: string;
  expire_at: string;
  total_invited: number;
  valid_invited: number;
  total_rewards: number;
}

export interface InvitedRiderInfo {
  rider_id: number;
  phone: string;
  nickname: string;
  status: number;
  register_at: string;
  first_order_at: string;
  task_completed_at: string;
  reward_amount: number;
}

export interface TaskInfo {
  task_id: number;
  task_name: string;
  task_desc: string;
  task_type: number;
  target_value: number;
  reward_amount: number;
  time_limit_days: number;
  start_time: string;
  end_time: string;
}

export interface TaskProgressInfo {
  task_id: number;
  task_name: string;
  current_value: number;
  target_value: number;
  progress_percent: number;
  status: number;
  deadline: string;
  is_claimed: boolean;
}

export interface RewardRecord {
  record_id: number;
  task_id: number;
  task_name: string;
  reward_amount: number;
  reward_type: number;
  status: number;
  created_at: string;
  issued_at: string;
}

export const referralApi = {
  generateInviteCode: async (riderId: number): Promise<{ invite_code_info: InviteCodeInfo }> => {
    const response = await apiClient.post('/rider/v1/referral/invite-code/generate', {
      rider_id: riderId,
    });
    return response.data;
  },

  getMyReferralInfo: async (riderId: number) => {
    const response = await apiClient.get('/rider/v1/referral/my-info', {
      params: { rider_id: riderId },
    });
    return response.data;
  },

  getInviteRecordList: async (riderId: number, page: number, pageSize: number, status?: number) => {
    const response = await apiClient.get('/rider/v1/referral/invite-records', {
      params: { rider_id: riderId, page, page_size: pageSize, status },
    });
    return response.data;
  },

  validateInviteCode: async (inviteCode: string) => {
    const response = await apiClient.post('/rider/v1/referral/invite-code/validate', {
      invite_code: inviteCode,
    });
    return response.data;
  },

  bindReferralRelation: async (newRiderId: number, inviteCode: string, phone: string) => {
    const response = await apiClient.post('/rider/v1/referral/bind', {
      new_rider_id: newRiderId,
      invite_code: inviteCode,
      phone,
    });
    return response.data;
  },

  getTaskList: async (riderId: number): Promise<{ tasks: TaskInfo[] }> => {
    const response = await apiClient.get('/rider/v1/referral/tasks', {
      params: { rider_id: riderId },
    });
    return response.data;
  },

  getTaskProgress: async (riderId: number, taskId: number): Promise<{ progress: TaskProgressInfo }> => {
    const response = await apiClient.get(`/rider/v1/referral/task-progress/${taskId}`, {
      params: { rider_id: riderId },
    });
    return response.data;
  },

  claimTaskReward: async (riderId: number, taskId: number) => {
    const response = await apiClient.post(`/rider/v1/referral/tasks/${taskId}/claim`, {
      rider_id: riderId,
    });
    return response.data;
  },

  getRewardRecordList: async (riderId: number, page: number, pageSize: number, status?: number) => {
    const response = await apiClient.get('/rider/v1/referral/reward-records', {
      params: { rider_id: riderId, page, page_size: pageSize, status },
    });
    return response.data;
  },
};

export const TASK_TYPE = {
  FIRST_ORDER: 1,
  ONLINE_TIME: 2,
  ORDER_COUNT: 3,
};

export const INVITED_RIDER_STATUS = {
  REGISTERED: 1,
  FIRST_ORDER: 2,
  TASK_COMPLETED: 3,
  REWARDED: 4,
};
